# Analytics in Practice

# Data Wrangling - Assignment 1 
- By Srushti Padade

# Summary

The libraries such as Pandas, Json, Pandas_profiling in Python plays an important role in data wrangling process of preprocessing. We also have libraries like dplyr, ISLR, etc in R language. Any real time data have various constraints, such as the data inconsistaency, missing values, duplicacy, or major issue with reading the data due to its format.
Thus, just loading the data does not necessarily help with descriptive statistics of the data, but some time it takes more extraction by manipulation, as we have done it with the json file. Cleaning of data is must to gain the information it store. Therefore, applying the statistics measures we can gain an insight of the data to build a model and further help with Machine learning.

Here while dealing with the 3 different data set of varied format we have learnt the techniques to read these files and summaries the data they have.The data can be of different datatype which can create a difficulty in manipulating them, but there are methods to be used to change and manipulate the data. We have also learnt the drawback of an raw data and also the cleaning techniques of them. Extracting the possible information also could be gathered for a better data visualization.


In [1]:
# The below libraries are imported for the data wrangling purpose. 
# Pandas is the renoumced library in python that helps with multiplefunctionalities.
# Similarly, Json libaray is used to manipulate the data from .json file format.

In [2]:
import pandas as pd
import json

#Reading the .csv data file
Companies_Data = pd.read_csv('us_companies.csv')
Agencies_Data = pd.read_csv('us_agencies.csv')

# Computes the size and dimension of the data.
CSize = Companies_Data.size
ASize = Agencies_Data.size

CShape = Companies_Data.shape
AShape = Agencies_Data.shape

In [3]:
print("The Size of Company Data is :", CSize)

print("The Size of Agencies Data is :", ASize)

print("The Dimensions of Company Data is :", CShape)

print("The Dimensions of Agencies Data is :", AShape)

The Size of Company Data is : 11638
The Size of Agencies Data is : 12353
The Dimensions of Company Data is : (529, 22)
The Dimensions of Agencies Data is : (1123, 11)


In [4]:
# To get all the Variables of the data set.
Companies_Data.columns

Index(['company_name_id', 'company_name', 'url', 'year_founded', 'city',
       'state', 'country', 'zip_code', 'full_time_employees', 'company_type',
       'company_category', 'revenue_source', 'business_model', 'social_impact',
       'description', 'description_short', 'source_count', 'data_types',
       'example_uses', 'data_impacts', 'financial_info', 'last_updated'],
      dtype='object')

In [5]:
Agencies_Data.columns

Index(['agency_name', 'agency_abbrev', 'agency_type', 'subagency_name',
       'subagency_abbrev', 'url', 'used_by', 'used_by_category', 'used_by_fte',
       'dataset_name', 'dataset_url'],
      dtype='object')

# Question 1: Are there any missing columns?

By looking at the Data columns we cannot determine the missing columns unless we have a DOMAIN knowldge about the data.
But simply looking the columns there do not seems to be any missing columns.

In [6]:
# To know the data type of each variable.
Agencies_Data.dtypes

agency_name         object
agency_abbrev       object
agency_type         object
subagency_name      object
subagency_abbrev    object
url                 object
used_by             object
used_by_category    object
used_by_fte         object
dataset_name        object
dataset_url         object
dtype: object

In [7]:
Companies_Data.dtypes

company_name_id         object
company_name            object
url                     object
year_founded           float64
city                    object
state                   object
country                 object
zip_code               float64
full_time_employees     object
company_type            object
company_category        object
revenue_source          object
business_model          object
social_impact           object
description             object
description_short       object
source_count            object
data_types              object
example_uses            object
data_impacts            object
financial_info          object
last_updated            object
dtype: object

# Question 2: Are there any missing column names or errors in the column names? If so, name those columns.

All the Columns are named in both the dataset. 
The column used_by_fte does look like a column that could have a better name.

In [8]:
# The Null values in each column can be computed as below.
Agencies_Data.isnull().sum()

agency_name           0
agency_abbrev       313
agency_type           0
subagency_name        0
subagency_abbrev    709
url                 292
used_by               0
used_by_category      4
used_by_fte          44
dataset_name        548
dataset_url         808
dtype: int64

In [9]:
Companies_Data.isnull().sum()

company_name_id          0
company_name             0
url                      0
year_founded             1
city                    33
state                    0
country                  0
zip_code                37
full_time_employees     29
company_type            16
company_category         3
revenue_source          10
business_model          76
social_impact          513
description              0
description_short        0
source_count           303
data_types             387
example_uses           521
data_impacts             0
financial_info         387
last_updated             0
dtype: int64

In [10]:
# Here we have the total NULL values from the data set.
Agencies_Data.isnull().sum().sum()

2718

In [11]:
Companies_Data.isnull().sum().sum()

2316

# Question 3: Are there any values in the columns missing?

There are many missing values in the dataset.
Above is the detail view of ll the missing values column wise and total missing values for Agencies data is 2718, whereas for companies data is 2316.

In [12]:
# To have a descriptive statistics or summary of the data below command is helpful.
Companies_Data.describe(include = 'all')

company_name_id    company_name              url  \
count                                    529             529              529   
unique                                   529             529              528   
top     aureus-sciences-now-part-of-elsevier  H3 Biomedicine  www.careset.com   
freq                                       1               1                2   
mean                                     NaN             NaN              NaN   
std                                      NaN             NaN              NaN   
min                                      NaN             NaN              NaN   
25%                                      NaN             NaN              NaN   
50%                                      NaN             NaN              NaN   
75%                                      NaN             NaN              NaN   
max                                      NaN             NaN              NaN   

        year_founded      city state country      zip_code  \
count     528.000000       496   529     529    492.000000   
unique           NaN       202    39       1           NaN   
top              NaN  New York    CA      us           NaN   
freq             NaN        83   132     529           NaN   
mean     1992.700758       NaN   NaN     NaN  47455.878049   
std        33.853259       NaN   NaN     NaN  37374.897009   
min      1799.000000       NaN   NaN     NaN      0.000000   
25%      1994.000000       NaN   NaN     NaN  10018.750000   
50%      2007.000000       NaN   NaN     NaN  37025.500000   
75%      2010.000000       NaN   NaN     NaN  94025.000000   
max      2015.000000       NaN   NaN     NaN  98144.000000   

       full_time_employees company_type             ...              \
count                  500          513             ...               
unique                   8            9             ...               
top                   1-10      Private             ...               
freq                   143          396             ...               
mean                   NaN          NaN             ...               
std                    NaN          NaN             ...               
min                    NaN          NaN             ...               
25%                    NaN          NaN             ...               
50%                    NaN          NaN             ...               
75%                    NaN          NaN             ...               
max                    NaN          NaN             ...               

              business_model      social_impact  \
count                    453                 16   
unique                    27                 11   
top     Business to Business  Healthcare access   
freq                     172                  3   
mean                     NaN                NaN   
std                      NaN                NaN   
min                      NaN                NaN   
25%                      NaN                NaN   
50%                      NaN                NaN   
75%                      NaN                NaN   
max                      NaN                NaN   

                                              description  \
count                                                 529   
unique                                                529   
top     J.P. Morgan Chase is a global financial servic...   
freq                                                    1   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                        description_short source_count  \
count                                 

In [13]:
Agencies_Data.describe(include = 'all')

agency_name agency_abbrev agency_type subagency_name  \
count                     1123           810        1123           1123   
unique                      70            60           5             56   
top     Department of Commerce          USDC     Federal        General   
freq                       194           194         760            707   

       subagency_abbrev                                              url  \
count               414                                              831   
unique               53                                              111   
top              Census  http://catalog.data.gov/organization/census-gov   
freq                 83                                               83   

                used_by used_by_category used_by_fte  dataset_name dataset_url  
count              1123             1119        1079           575         315  
unique              526               20           8           506         276  
top     FindTheBest.com  Data/Technology        1-10  zipcode data  census.gov  
freq                 30              190         392            12           5

# Question 4: How is data organized in each column? Is it properly organized?

The summary statistics of the data have a good picture how the data is organised.

For Agencies the data is a catagorical value and hence the distribution cannot be obtained.
However, for Companies data the distribution is available.

The data organised is well manner in the columns but due to missing values it is difficult to extract the ditribution of the data. Also the data_impact column is not organized as it contains a empty lists in it.

# Question 5: Is data in the proper shape for further analysis? If not, why? Explain.

The data are not in the proper shape for further analysis.
- There are lots of missing values in the dataset(73%).
- There is also redandancy in the data.
- The is made complexed by direct storage of the list values in column like data_impact.
- Few columns are having highly corelated data which might affect the training later. 

In [30]:
# Cleaning the dataset.
Clean_Companies = Companies_Data.dropna(axis='columns')

Clean_Companies.head()

company_name_id          company_name                          url  \
0  3-round-stones-inc  3 Round Stones, Inc.      http://3RoundStones.com   
1    48-factoring-inc     48 Factoring Inc.  https://www.48factoring.com   
2         5psolutions           5PSolutions          www.5psolutions.com   
3      abt-associates        Abt Associates                 abtassoc.com   
4              accela                Accela        http://www.accela.com   

  state country                                        description  \
0    DC      us  3 Round Stones produces a platform for publish...   
1    PA      us  The company mission is to provide finance to s...   
2    VA      us  At 5PSolutions, we wish to make all basic info...   
3    MA      us  Abt Associates is a mission-driven, internatio...   
4    CA      us  Accela powers thousands of services and millio...   

                                   description_short  \
0  Our Open Source platform is used by the Fortun...   
1  48 Factoring Inc. is one of the best financial...   
2      5PSolutions are artisans of mobile platforms.   
3  Abt Associates is a mission-driven, global lea...   
4  Accela improves citizen engagement by making i...   

                                        data_impacts  \
0                                                 []   
1  [u'Cost efficiency', u'Job growth', u'Revenue ...   
2                                                 []   
3                                                 []   
4                                                 []   

                 last_updated  
0  2014-11-12 14:44:25.969871  
1  2015-05-18 11:36:39.902880  
2  2014-11-12 15:09:35.485949  
3  2014-09-11 17:23:21.390289  
4  2014-09-16 16:30:49.172802

In [31]:
Clean_Agencies = Agencies_Data.dropna()

Clean_Agencies.head()

agency_name agency_abbrev agency_type              subagency_name  \
6   Department of Labor           DOL     Federal  Bureau of Labor Statistics   
7   Department of Labor           DOL     Federal  Bureau of Labor Statistics   
8   Department of Labor           DOL     Federal  Bureau of Labor Statistics   
9   Department of Labor           DOL     Federal  Bureau of Labor Statistics   
10  Department of Labor           DOL     Federal  Bureau of Labor Statistics   

   subagency_abbrev                                           url  \
6               BLS  http://catalog.data.gov/organization/bls-gov   
7               BLS  http://catalog.data.gov/organization/bls-gov   
8               BLS  http://catalog.data.gov/organization/bls-gov   
9               BLS  http://catalog.data.gov/organization/bls-gov   
10              BLS  http://catalog.data.gov/organization/bls-gov   

              used_by      used_by_category used_by_fte dataset_name  \
6   Fuzion Apps, Inc.  Lifestyle & Consumer        1-10        State   
7   Fuzion Apps, Inc.  Lifestyle & Consumer        1-10     National   
8        Investormill  Finance & Investment        1-10          CPI   
9        Investormill  Finance & Investment        1-10          PPI   
10       Investormill  Finance & Investment        1-10   Employment   

                            dataset_url  
6     http://www.bls.gov/oes/oes_dl.htm  
7     http://www.bls.gov/oes/oes_dl.htm  
8       http://www.bls.gov/cpi/data.htm  
9       http://www.bls.gov/data/#prices  
10  http://www.bls.gov/data/#employment

# Question 6: How will you fix this dataset? Describe the methods you will use to fix this dataset for further analysis?

The Data set could be cleaned using below measures:
- Remove the missing values from the column.
- The columns could be droped with very few data entries.
- The redanduncy can be eliminated be removing corelated columns. (Remove duplicated)
- The columns with the list values can be extracted to generate new columns for cleaner visibility.

# Question 7: How are the two datasets linked to each other? Is there a common “primary key” to connect the two datasets?

The two dataset are linked to each other using the company name: 
Comapnies_Data:company_name = Agencies_Data:used_by. 

In [16]:
# Loading the .json file in the environment.

In [17]:
with open('ChicagoTraffic.json') as f:
    data = json.load(f)

In [18]:
# Here the json file is stored with 2 key roots - MetaData and Data we have to extract the data individually.
DataEntries = pd.Series(data["data"])

MetaData = pd.Series(data["meta"]["view"]["columns"])

In [19]:
# Here we can see all the summary of each column from the data by manipulating the Metadata.
pd.DataFrame.from_dict(MetaData.to_dict())

0          1          2            3   \
cachedContents             NaN        NaN        NaN          NaN   
computationStrategy        NaN        NaN        NaN          NaN   
dataTypeName         meta_data  meta_data  meta_data    meta_data   
fieldName                 :sid        :id  :position  :created_at   
flags                 [hidden]   [hidden]   [hidden]     [hidden]   
format                      {}         {}         {}           {}   
id                          -1         -1         -1           -1   
name                       sid         id   position   created_at   
position                     0          0          0            0   
renderTypeName       meta_data  meta_data  meta_data    meta_data   
subColumnTypes             NaN        NaN        NaN          NaN   
tableColumnId              NaN        NaN        NaN          NaN   
width                      NaN        NaN        NaN          NaN   

                                4            5              6          7   \
cachedContents                 NaN          NaN            NaN        NaN   
computationStrategy            NaN          NaN            NaN        NaN   
dataTypeName             meta_data    meta_data      meta_data  meta_data   
fieldName            :created_meta  :updated_at  :updated_meta      :meta   
flags                     [hidden]     [hidden]       [hidden]   [hidden]   
format                          {}           {}             {}         {}   
id                              -1           -1             -1         -1   
name                  created_meta   updated_at   updated_meta       meta   
position                         0            0              0          0   
renderTypeName           meta_data    meta_data      meta_data  meta_data   
subColumnTypes                 NaN          NaN            NaN        NaN   
tableColumnId                  NaN          NaN            NaN        NaN   
width                          NaN          NaN            NaN        NaN   

                                                                    8   \
cachedContents       {'non_null': 1279, 'average': '689.93745113369...   
computationStrategy                                                NaN   
dataTypeName                                                    number   
fieldName                                                           id   
flags                                                              NaN   
format                                            {'aggregate': 'sum'}   
id                                                             2459455   
name                                                               ID    
position                                                             1   
renderTypeName                                                  number   
subColumnTypes                                                     NaN   
tableColumnId                                                  1358385   
width                                                               50   

                                                                    9   \
cachedContents       {'non_null': 1279, 'largest': '996 South', 'nu...   
computationStrategy                                                NaN   
dataTypeName                                                      text   
fieldName                        traffic_volume_count_location_address   
flags                                                              NaN   
format                                                              {}   
id                                                             2459456   
name                            Traffic Volume Count Location  Address   
position                                                             2   
renderTypeName                                                    text   
subColumnTypes                                                     NaN   
tableColumnId                                            

In [20]:
# The below librar provides a better view of the data with respect to its descriptive statistics.

#Installation
#conda install -c conda-forge pandas-profiling

#import pandas_profiling
#from pandas_profiling import ProfileReport

#profile = ProfileReport(Companies_data)
#profile

#This libarary gives a clearer picture of the data with all its distribution like: 
#Percentage missing value, catagorical data, distribution of the data, etc.

# Question 1: How many variables are in the dataset?

There are 23 variables in the data set.

In [21]:
pd.DataFrame.from_dict(MetaData.to_dict()).index

Index(['cachedContents', 'computationStrategy', 'dataTypeName', 'fieldName',
       'flags', 'format', 'id', 'name', 'position', 'renderTypeName',
       'subColumnTypes', 'tableColumnId', 'width'],
      dtype='object')

# Question 2: Name all the variables?

Above are the names of all the variables from the dataset.

In [22]:
# To know the traffic of the streets we have stored he street names as a list for computation.
Traffic = list()

Street = ["100th St", "101 St", "102 St", "103rd St", "104th St", "105th St", "106th St", "107th St", "108th St",
       "109th St", "110th St", "111th St", "112th St", "113th St", "114th St", "115th St"]

In [23]:
#The below loop iterates over the Data entries and retrieve the Traffic count of each street.

In [24]:
for item in DataEntries:
    if item[10] in Street:
        temp = int(item[12])
        Traffic.append(temp)

# Question 3: What is the total traffic of vehicles on 100th street to 115th street?

The traffic count is computed below.

In [25]:
#The totl traffic is the sum of all the individual traffic.
print("The Total traffic between street 100th to 115 is", sum(Traffic))

The Total traffic between street 100th to 115 is 443200


In [26]:
#Similarly, for the Coordinates of the location we can reterive the traffic count.
Coordinates = list()

In [27]:
for item in DataEntries:
    if item[14] == "41.66836" and item[15] == "-87.620176":
        temp = int(item[12])
    elif item[14] == "41.651861" and item[15] == "-87.54501":
        temp1 = int(item[12])
        

# Question 4: What is the total traffic of vehicles on geolocations, (41.651861, -87.54501) and (41.66836, -87.620176)

The total traffic for the given geographic location is computed below.

In [28]:
print("Total traffic with at Latitute = 41.66836 and longitude = -87.620176 is", temp)

print("Total traffic with at Latitute = 41.651861 and longitude = -87.54501 is", temp1)

print("Total traffic at the given coordinates will be", temp+temp1)

Total traffic with at Latitute = 41.66836 and longitude = -87.620176 is 8900
Total traffic with at Latitute = 41.651861 and longitude = -87.54501 is 4700
Total traffic at the given coordinates will be 13600
